### PR TITLE
[Samples] Add samples for testing operations in Q#

### PR DESCRIPTION
**Pull request by Manvi-Agrawal**
Fixes part of #1268 
- Added samples for testing operations using `dump_operation`. 
- Also, using `CheckOperationsAreEqual` inside Fact function to demonstrate another way

- Hopefully, the new testing samples should be executed in CI. @tcNickolas could you please start a CI run. Snapshot from locally running `build.py`. 
image

- Running in CI:
image

### CODE REVIEW

File: samples/testing/dump_operation/README.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
+**Copyright (c) Microsoft Corporation. All rights reserved.**  
+**Licensed under the MIT License.**
+
+# Testing Operations in Modern QDK
+This sample project demonstrates different approaches to testing operations in Modern QDK, which unlike Classic QDK, doesn't support native Q# tests.
```
Comment (swernli) [inline samples/testing/dump_operation/README.md:5]: So far we've avoided specific references to the Classic QDK in our guidance here, since this QDK is expected to stand on it's own and not require folks to be familiar with the Classic QDK. I'd rephrase this as:
```suggestion
# Testing Operations in the QDK
This sample project demonstrates different approaches to testing operations in the QDK, both via Python and within Q# itself.
```

File: samples/testing/dump_operation/README.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+## Testing Methods
+
+There are two primary ways to test operations in Modern QDK:
```
Comment (swernli) [inline samples/testing/dump_operation/README.md:9]: Same here, just refer to it as the QDK.
```suggestion
There are two primary ways to test operations in the QDK:
```

File: samples/testing/dump_operation/README.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+## Installation
+- Install the `qsharp` python package by following the instructions mentioned [here](https://learn.microsoft.com/azure/quantum/install-overview-qdk#add-support-for-python-and-jupyter-notebooks).
+- Install `pytest` python package.
+
+## Running the sample
+Open the `samples/testing/dump_operation` directory, and run `pytest` command.
+
+## Reference Links:
+- [Q# Testing guide](https://learn.microsoft.com/azure/quantum/user-guide/testing-debugging).
+- [Getting started with modern QDK](https://learn.microsoft.com/azure/quantum/install-overview-qdk)
```
Comment (swernli) [inline samples/testing/dump_operation/README.md:36]: Here too, this can just be the QDK.
```suggestion
- [Getting started with the QDK](https://learn.microsoft.com/azure/quantum/install-overview-qdk)
```

File: samples/testing/dump_operation/README.md (hunk 1 above)
Comment (swernli) [reply samples/testing/dump_operation/README.md]: I don't think that is needed, the guidance you have put together here should be enough for anyone searching for information on testing.

File: samples/testing/dump_operation/src/SWAP.qs [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    }
+
+    // # Summary
+    /// Operation for testing with `TestEquivalence` operation.
+    ///
+    /// # Input
+    /// ## qs
+    /// Input qubit register
+    operation ApplySWAP2(qs : Qubit[]) : Unit is Ctl + Adj {
+        SWAP(qs[0], qs[1]);
```
Comment (swernli) [inline samples/testing/dump_operation/src/SWAP.qs:33]: Perhaps it would be worth having the second SWAP implementation be the CNOT based decomposition? I'm not sure if this overcomplicates or distracts from the goal of the sample, so I'm fine if it stays the same for simplicity.

File: samples/testing/dump_operation/src/SWAP.qs [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+///
+/// # Description
+/// Organizing code into multiple Q# source files is an important part of
+/// writing readable and maintainable code. In this project, we have `BellState.qs`, `SWAP.qs`,
+/// and `OperationEquivalence.qs`, which contain the operation to be tested.
+/// The presence of a Q# manifest file (`qsharp.json`) tells the compiler
+/// to include all Q# files under `src/`.
+/// These will be tested by Python wrapper, `test_dump_operation.py`
+
+namespace SWAP {
```
Comment (swernli) [inline samples/testing/dump_operation/src/SWAP.qs:15]: This is a bit of a nitpick, but it can be a little confusing when the namespace has the same name as an intrinsic gate. Maybe `Test_SWAP` could avoid any conceptual overlap with the operation `SWAP`?

**Review by swernli — APPROVED** [review]
A few more minor comments, but otherwise looks good!

File: build.py [hunk 6]
```diff
@@ -363,6 +363,43 @@ def run_python_integration_tests(cwd, interpreter):
     )
     step_end()
 
+if build_samples:
```
Comment (swernli) [inline build.py:366]: As of https://github.com/microsoft/qsharp/pull/1739 all samples testing now occurs via `cargo test` and auto-generated Rust unit tests, so this section was deleted. I think the automatic merge may have reintroduced it. You can delete it here and confirm the added project is automatically picked up and compiled as part of the Rust tests.

File: samples/testing/dump_operation/README.md [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+# Testing Operations in the QDK
+This sample project demonstrates different approaches to testing operations in the QDK, both via Python and within Q# itself.
+
+## Testing Methods
+
+There are two primary ways to test operations in the QDK:
+
+1. **Dumping Operation Output:**
+   - Use the `dump_operation` Python API to retrieve the operation's representation and compare it against the expected output.
```
Comment (tcNickolas) [inline samples/testing/dump_operation/README.md:12]: Technically quantum operations don't have output, so this phrasing is confusing. I'd clarify that it's matrix representation instead
```suggestion
1. **Using Operation Matrix Representation:**
   - Use the `dump_operation` Python API to retrieve the operation's representation as a matrix and compare it against the expected matrix.
```

File: samples/testing/dump_operation/README.md [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+2. **Q# `Fact` Assertions:**
+   - Define a `Fact` function in your Q# code that uses the `CheckOperationsAreEqual` operation to verify if two operations are identical. The `Fact` function asserts that the check returns `true`.
```
Comment (tcNickolas) [inline samples/testing/dump_operation/README.md:15]:
```suggestion
   - Use a `Fact` function in your Q# code that uses the `CheckOperationsAreEqual` operation to verify if two operations are identical. The `Fact` function asserts that the check returns `true`.
```

File: samples/testing/dump_operation/README.md [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+## Project Structure
+This sample project is a multi-file Q# project that showcases both testing methods. The project structure is as follows:
+
+- src
+    - "BellState.qs": Q# file containing the `AllBellStates` operation to be tested
+    - "Test_SWAP.qs": Q# file containing the `ApplySWAP1` and `ApplySWAP2` operations to be tested
+    - "OperationEquivalence.qs": Q# file containing the `TestEquivalence` operation to be called in python wrapper
+- "qsharp.json": Q# project configuration file
+- test_dump_operation.py: "Python wrapper containing tests"
```
Comment (tcNickolas) [inline samples/testing/dump_operation/README.md:25]: This list uses inconsistent formatting for file names (using quotes or not) and text (last line has the next explanation in quotes). Let's unify; I don't think we need quotes for file names, if we want to highlight them, ` works better, and we don't need quotes around text

File: samples/testing/dump_operation/README.md [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+## Installation
+- Install the `qsharp` python package by following the instructions mentioned [here](https://learn.microsoft.com/azure/quantum/install-overview-qdk#add-support-for-python-and-jupyter-notebooks).
+- Install `pytest` python package.
+
+## Running the sample
+Open the `samples/testing/dump_operation` directory, and run `pytest` command.
```
Comment (tcNickolas) [inline samples/testing/dump_operation/README.md:32]: The directory name emphasizes one of the operations used in testing, but since the project has grown to show two approaches, I would rename it to emphasize the kind of thing we test instead - call it just "operations", so that the whole path says "testing operations"

File: samples/testing/dump_operation/src/Test_SWAP.qs [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+///
+/// # Description
+/// Organizing code into multiple Q# source files is an important part of
+/// writing readable and maintainable code. In this project, we have `BellState.qs`, `Test_SWAP.qs`,
+/// and `OperationEquivalence.qs`, which contain the operation to be tested.
+/// The presence of a Q# manifest file (`qsharp.json`) tells the compiler
+/// to include all Q# files under `src/`.
+/// These will be tested by Python wrapper, `test_dump_operation.py`
+
+namespace Test_SWAP {
```
Comment (tcNickolas) [inline samples/testing/dump_operation/src/Test_SWAP.qs:15]: This file contains the main code that is tested rather than the test code, so I would not call the namespace and the file starting with Test. It's really confusing that it uses Test in the name and the file that has the test logic does not!

File: samples/testing/dump_operation/test_dump_operation.py [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+def test_single_qubit_not_gate() -> None:
+    qsharp.init(target_profile=qsharp.TargetProfile.Unrestricted)
+
+    res = dump_operation("qs => X(qs[0])", 1)
+    assert res == [
+        [complex(0.0, 0.0), complex(1.0, 0.0)],
+        [complex(1.0, 0.0), complex(0.0, 0.0)],
+    ]
+
+def test_single_qubit_superposition() -> None:
```
Comment (tcNickolas) [inline samples/testing/dump_operation/test_dump_operation.py:26]: This test is named differently from the others, I'd suggest unified naming scheme
```suggestion
def test_single_qubit_hadamard_gate() -> None:
```

File: samples/testing/dump_operation/src/OperationEquivalence.qs [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+
+namespace OperationEquivalence {
+    open Microsoft.Quantum.Diagnostics;
+    open Test_SWAP;
+    /// # Summary
+    /// Verifies the equivalence of quantum operations using `Fact` function
+    /// and the `CheckOperationsAreEqual` operation.
+    operation TestEquivalence() : Unit {
+        let actual = qs => Test_SWAP.ApplySWAP1(qs);
+        let expected = qs => Test_SWAP.ApplySWAP2(qs);
```
Comment (tcNickolas) [inline samples/testing/dump_operation/src/OperationEquivalence.qs:25]: You shouldn't need lambdas here, since you're not defining new operations using partial application or applying the operations to a part of an array, just saying actual = ApplySWAP1 should work

File: samples/testing/dump_operation/src/OperationEquivalence.qs [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+        Fact(CheckOperationsAreEqual(2, actual, expected) == true, "Actual and expected operation should be same");
```
Comment (tcNickolas) [inline samples/testing/dump_operation/src/OperationEquivalence.qs:26]: You don't need to compare a Boolean to true to convert it to a Boolean

File: samples/testing/dump_operation/src/OperationEquivalence.qs [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+///
+/// # Description
+/// This code builds upon the concepts explained in the README file,
+/// demonstrating how to organize Q# code into multiple files for testing.
+/// Here, we have separate files (`BellState.qs`, `Test_SWAP.qs`, and `OperationEquivalence.qs`)
+/// containing individual operations under test.
+
+/// The presence of a Q# manifest file (`qsharp.json`) instructs the compiler
+/// to include all Q# files within the `src` directory. These operations are tested
+/// by the Python wrapper script, `test_dump_operation.py`.
```
Comment (tcNickolas) [inline samples/testing/dump_operation/src/OperationEquivalence.qs:15]: For a multi-file project, I would keep the description of the project file structure to the README file, and use summary of the specific files within it to describe just the content of that file. This way, there is no information duplication, the files look cleaner, and it's easier to update them if we rename a file or add an extra one

File: samples/testing/dump_operation/test_dump_operation.py [hunk 16]
```diff
@@ -0,0 +1,119 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (tcNickolas) [inline samples/testing/dump_operation/test_dump_operation.py:1]: To match the project renaming suggestion, I'd rename this file to "test_operations.py", since it showcases both approaches

File: samples/testing/dump_operation/test_dump_operation.py [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+# Licensed under the MIT License.
+
+import qsharp
+from qsharp.utils import dump_operation
+
+def test_empty_operation() -> None:
+    qsharp.init(target_profile=qsharp.TargetProfile.Unrestricted)
```
Comment (tcNickolas) [inline samples/testing/dump_operation/test_dump_operation.py:8]: Do we need to set this profile explicitly every time? I think it's the default setting, my tests get away with not using it, just saying qsharp.init()

File: samples/testing/dump_operation/test_dump_operation.py [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+    
+    res = dump_operation("qs => ()", 1)
+    assert res == [
+        [complex(1.0, 0.0), complex(0.0, 0.0)],
```
Comment (tcNickolas) [inline samples/testing/dump_operation/test_dump_operation.py:12]: Do you need to call complex explicitly, or can you compare the numbers with just [1, 0], [0, 1]? I think Python is nice enough about type casting to allow to compare floating point with complex directly, I do that in my tests

**Review by tcNickolas — CHANGES_REQUESTED** [review]
Looks good! I like the variety of tests you have, including the ones showing partial application to test operations with parameters!

I left suggestions for code and docs readability.

File: samples/testing/operations/src/Test_SWAP.qs [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+// Licensed under the MIT License.
+
+/// # Sample
+/// Multi File Testing Project
+///
+/// # Description
+/// This code builds upon the concepts explained in the README file,
+/// demonstrating how to organize Q# code into multiple files for testing.
+
+namespace Test_SWAP {
```
Comment (tcNickolas) [inline samples/testing/operations/src/Test_SWAP.qs:11]: The previous comment chain disappeared, so here's a new one: I agree that it's good to name namespaces with names distinct from the operations defined in them, but it's also good not to put non-test logic in namespace named "test". We can use "CustomOperation" or "CustomSWAP" to satisfy both

**Comment by swernli** [comment]
Thanks for merging main and confirming things work well with the new sample testing. I can see the auto-generated tests run successfully as well as the new tests added here. It looks great!

File: samples/testing/operations/test_operations.py [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+import qsharp
+from qsharp.utils import dump_operation
+
+@pytest.fixture(autouse=True)
+def setup():
+    """Fixture to execute asserts before a test is run"""
+    # You can set the target profile here if you want
+    # By default, it uses "Unrestricted" profile
+    qsharp.init(target_profile=qsharp.TargetProfile.Base)
+    yield # this is where the testing happens
```
Comment (tcNickolas) [inline samples/testing/operations/test_operations.py:14]: Several questions on this piece:
* The doc comment mentions asserts but the code doesn't actually include asserts
* I understand that this executes qsharp.init before every test, correct? But all tests except the first one call qsharp.init anyways, re-initializing qsharp again and not benefitting from this fixture?
* You shouldn't need to specify profile at all, the tests can run without that, and here you want to have the minimal test code that achieves the goal.
* I would swap the init argument for `project_root='.'`, since you do actually need that for the later tests and they don't work without it. Right now, the later tests call init three times, one from setup, one with setting target profile, and one with setting project root, and I think only the last one is actually used

File: samples/testing/operations/src/BellState.qs [hunk 21]
```diff
@@ -0,0 +1,42 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+/// # Sample
+/// Multi File Testing Project (Refer to README for Project Overview)
+///
+/// # Description
+/// This code builds upon the concepts explained in the README file,
+/// demonstrating how to organize Q# code into multiple files for testing.
+
```
Comment (tcNickolas) [inline samples/testing/operations/src/BellState.qs:10]: We do not need to repeat the reference to the README file in every Q# file. A README file in the root folder implies that it has the descriptions that are relevant to all files, and each file needs to include only its own documentation. Please remove this summary from this file and the other two Q# files in this folder

File: samples/testing/operations/src/BellState.qs [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+    /// # Summary
+    /// Operation that generates all bell states for testing with `dump_operation.py`.
+    ///
+    /// # Input
+    /// ## qs
+    /// Input qubit register
+    ///
+    /// ## choice
+    /// Bell state to construct.
+    /// 0: |Φ+〉(PhiPlus)
```
Comment (tcNickolas) [inline samples/testing/operations/src/BellState.qs:21]: I think 〉 is the ket symbol that looks weird in some environments (https://github.com/microsoft/QuantumKatas/issues/4 was actually one of the first issues on the original katas!), and we try to use ⟩ instead.

If you want to go all the way to fancy with Unicode ket notations, Unicode has subscripts for + and - (https://en.wikipedia.org/wiki/Unicode_subscripts_and_superscripts), Bell state notations look better with them, since regular plus and minus look a bit like tensor products of |Φ⟩ with |+⟩ and |-⟩

File: samples/testing/operations/src/OperationEquivalence.qs [hunk 23]
```diff
@@ -0,0 +1,24 @@
...
+///
+/// # Description
+/// This code builds upon the concepts explained in the README file,
+/// demonstrating how to organize Q# code into multiple files for testing.
+
+namespace OperationEquivalence {
+    open Microsoft.Quantum.Diagnostics;
+    open CustomOperation;
+    /// # Summary
+    /// Verifies the equivalence of quantum operations using `Fact` function
```
Comment (tcNickolas) [inline samples/testing/operations/src/OperationEquivalence.qs:15]: Here, we need to mention that this check compares two operations up to a global phase. This is important, for example, if the user tries to hunt down a global phase issue. The other approach to testing doesn't have this issue

File: samples/testing/operations/README.md [hunk 24]
```diff
@@ -0,0 +1,34 @@
...
+
+2. **Q# `Fact` Assertions:**
+   - Use a `Fact` function in your Q# code that uses the `CheckOperationsAreEqual` operation to verify if two operations are identical. The `Fact` function asserts that the check returns `true`.
+
+## Project Structure
+This sample project is a multi-file Q# project that showcases both testing methods. The project structure is as follows:
+
+- src
+    - `BellState.qs`: Q# file containing the `AllBellStates` operation to be tested
+    - `Test_SWAP.qs`: Q# file containing the `ApplySWAP1` and `ApplySWAP2` operations to be tested
```
Comment (tcNickolas) [inline samples/testing/operations/README.md:19]: Need to update doc to match the code

File: samples/testing/operations/README.md [hunk 25]
```diff
@@ -0,0 +1,34 @@
...
+
+## Testing Methods
+
+There are two primary ways to test operations in the QDK:
+
+1. **Using Operation Matrix Representation:**
+   - Use the `dump_operation` Python API to retrieve the operation's representation as a matrix and compare it against the expected matrix.
+
+2. **Q# `Fact` Assertions:**
+   - Use a `Fact` function in your Q# code that uses the `CheckOperationsAreEqual` operation to verify if two operations are identical. The `Fact` function asserts that the check returns `true`.
```
Comment (tcNickolas) [inline samples/testing/operations/README.md:12]: Need to mention the "up to the global phase" part

File: samples/testing/operations/README.md [hunk 26]
```diff
@@ -0,0 +1,34 @@
...
+
+## Project Structure
+This sample project is a multi-file Q# project that showcases both testing methods. The project structure is as follows:
+
+- src
+    - `BellState.qs`: Q# file containing the `AllBellStates` operation to be tested
+    - `Test_SWAP.qs`: Q# file containing the `ApplySWAP1` and `ApplySWAP2` operations to be tested
+    - `OperationEquivalence.qs`: Q# file containing the `TestEquivalence` operation to be called in python wrapper
```
Comment (tcNickolas) [inline samples/testing/operations/README.md:20]: Python should be capitalized when used in a sentence (as opposed to python command line)

**Review by tcNickolas — CHANGES_REQUESTED** [review]
This looks a lot better, I like how the code got more concise and matrices in the test are human-readable!

I left some more comments, with them fixed this will be good to go!

File: samples/testing/operations/src/BellState.qs (hunk 22 above)
Comment (tcNickolas) [reply samples/testing/operations/src/BellState.qs:14]: That sample is actually a great example of why + and - in Bell state variables should be subscripts rather than regular symbols - it uses |Φ+⟩ (not a tensor product) side by side with |+0〉 (tensor product).

I'm happy to table this discussion for now, though. I've started an internal discussion about unifying Unicode in the samples

File: samples/testing/operations/README.md [hunk 27]
```diff
@@ -9,21 +9,21 @@ There are two primary ways to test operations in the QDK:
    - Use the `dump_operation` Python API to retrieve the operation's representation as a matrix and compare it against the expected matrix.
 
 2. **Q# `Fact` Assertions:**
-   - Use a `Fact` function in your Q# code that uses the `CheckOperationsAreEqual` operation to verify if two operations are identical. The `Fact` function asserts that the check returns `true`.
+   - Use a `Fact` function in your Q# code that uses the `CheckOperationsAreEqual` operation to verify if two operations are identical upto a global phase. The `Fact` function asserts that the check returns `true`.
```
Comment (tcNickolas) [inline samples/testing/operations/README.md:12]:
```suggestion
   - Use a `Fact` function in your Q# code that uses the `CheckOperationsAreEqual` operation to verify if two operations are identical up to a global phase. The `Fact` function asserts that the check returns `true`.
```

File: samples/testing/operations/README.md [hunk 28]
```diff
@@ -9,21 +9,21 @@ There are two primary ways to test operations in the QDK:
...
 
 ## Project Structure
 This sample project is a multi-file Q# project that showcases both testing methods. The project structure is as follows:
 
 - src
     - `BellState.qs`: Q# file containing the `AllBellStates` operation to be tested
-    - `Test_SWAP.qs`: Q# file containing the `ApplySWAP1` and `ApplySWAP2` operations to be tested
-    - `OperationEquivalence.qs`: Q# file containing the `TestEquivalence` operation to be called in python wrapper
+    - `CustomOperation.qs`: Q# file containing the `ApplySWAP1` and `ApplySWAP2` operations to be tested
```
Comment (tcNickolas) [inline samples/testing/operations/README.md:19]:
```suggestion
    - `CustomOperation.qs`: Q# file containing the `ApplySWAP` operation to be tested
```

File: samples/testing/operations/src/OperationEquivalence.qs [hunk 29]
```diff
@@ -1,20 +1,13 @@
...
-/// # Description
-/// This code builds upon the concepts explained in the README file,
-/// demonstrating how to organize Q# code into multiple files for testing.
-
 namespace OperationEquivalence {
     open Microsoft.Quantum.Diagnostics;
     open CustomOperation;
     /// # Summary
-    /// Verifies the equivalence of quantum operations using `Fact` function
+    /// Verifies the equivalence of quantum operations upto a global phase using `Fact` function
```
Comment (tcNickolas) [inline samples/testing/operations/src/OperationEquivalence.qs:8]:
```suggestion
    /// Verifies the equivalence of quantum operations up to a global phase using `Fact` function
```

File: samples/testing/operations/test_operations.py [hunk 30]
```diff
@@ -66,7 +57,6 @@ def test_custom_operation() -> None:
     ]
 
 def test_operation_no_args_in_qsharp_file() -> None:
-    qsharp.init(target_profile=qsharp.TargetProfile.Unrestricted)
     qsharp.init(project_root='.')
```
Comment (tcNickolas) [inline samples/testing/operations/test_operations.py:60]: Same comment, since we have a fixture that does exactly the same thing, we don't need to repeat .init in the individual tests. Please remove duplicate inits

**Review by tcNickolas — CHANGES_REQUESTED** [review]
Looks good. Hopefully one last round of suggestions!

Thank you!

**Review by tcNickolas — APPROVED** [review]
Looks great! Thank you so much for working on this!
